File: src/system/mips/cop0.js
import { PROCESSOR_ID, Exceptions } from "./consts";
import Exception from "./exception";

// NOTE: I use a system random number generator instead of the clock counter
// as this approach is faster and users should NOT depend on specific implementation
// for how this works

// NOTE: COP0 should be using the enable flags in _status, not test the kernel flag

const STATUS_CU3 = 0x80000000;
const STATUS_CU2 = 0x40000000;
const STATUS_CU1 = 0x20000000;
const STATUS_CU0 = 0x10000000;

const STATUS_RE	 = 0x02000000;

const STATUS_BEV = 0x00400000;
const STATUS_TS	 = 0x00200000;
const STATUS_PE	 = 0x00100000;
const STATUS_CM	 = 0x00080000;
const STATUS_PZ	 = 0x00040000;
const STATUS_SwC = 0x00020000;
const STATUS_IsC = 0x00010000;

const STATUS_IM	 = 0x0000FF00;

const STATUS_KUo = 0x00000020;
const STATUS_IEo = 0x00000010;
const STATUS_KUp = 0x00000008;
const STATUS_IEp = 0x00000004;
const STATUS_KUc = 0x00000002;
const STATUS_IEc = 0x00000001;

function random() {
	return (Math.random() * 0x37 + 0x8) & 0x3F;
}

export default class {
	resetCOP0() {
		// Status values
		this._status = STATUS_KUc | STATUS_BEV;
		this._cause = 0;
		this._epc = 0;

		// TLB
		this._entryLo = 0;
		this._entryHi = 0;
		this._badAddr = 0;
		this._targetAddr = 0;
		this._ptBase = 0;
		this._index	= 0;

		this._tlbLo = new Uint32Array(64);
		this._tlbHi = new Uint32Array(64);
		this._tlb   = [];
	}

	_copEnabled(cop) {
		return (((this._status >> 18) | ((this._status & STATUS_KUc) ? 1 : 0)) >>> cop) & 1;
	}

	_mfc0(reg, pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		switch (reg) {
		// Random registers
		case 0xF: // c0_prid
			return PROCESSOR_ID;

		// Virtual-memory registers
		case 0x0: // c0_index
			return this._index;
		case 0x1: // c0_random (non-deterministic, cheap method)
			return random() << 8;
		case 0x2: // c0_entrylo
			return this._entryLo;
		case 0xA: // c0_entryhi
			return this._entryHi;
		case 0x4: // c0_context
			return this._ptBase |
				((this._badAddr & 0xFFFFF800) >> 11);
		case 0x8: // c0_vaddr
			return this._badAddr;

		// Status/Exception registers
		case 0xC: // c0_status
			return this._status;
		case 0xD: // c0_cause
			return this._cause;
		case 0xE: // c0_epc
			return this._epc;
		default:
			return 0;
		}
	}

	_mtc0(reg, word, pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		switch (reg) {
		// Virtual-memory registers
		case 0x0: // c0_index
			this._index = word & 0x3F00;
			break ;
		case 0x2: // c0_entrylo
			this._entryLo = word & 0xFFFFFFC0;
			break ;
		case 0xA: // c0_entryhi
			this._entryHi = word & 0xFFFFFF00;
			break ;
		case 0x4: // c0_context
			this._ptBase = word & 0xFFE00000;
			break ;

		// Status/Exception registers
		case 0xC: // c0_status
		case 0xD: // c0_cause
			// TODO
		}
	}

	_translate(address, write) {
		let cached = true;
		if (address & 0x8000000 && ~this._status & STATUS_KUc) {
			throw Const.Exceptions.CoprocessorUnusable;
		}

		if (address & 0xE0000000 === 0xA0000000) {
			cached = false;
		}

		if (address & 0xC0000000 !== 0x80000000) {
			let page = address & 0xFFFFF000;
			let result = this._tlb[page | (this._entryHi & 0xFC0)] || this._tlb[page | 0xFFF];

			cached = ~result & 0x0800;
			this._targetAddr = address; // Holding register, trap will copy this to _badAddr if something goes wrong

			if (~result & 0x0200) {
				throw write ? Const.Exceptions.TLBStore : Const.Exceptions.TLBLoad;
			}

			if (~result & 0x0400 && write) {
				throw Const.Exceptions.TLBMod;
			}

			return (result & 0xFFFFF000) | (address & 0x00000FFF);
		} else {
			return address & 0x1FFFFFFC;
		}
	}

	_writeTLB(index) {
		// Ignore TLB entries that can cause a collision (normally would cause a system reset)
		if (this._tbl[this._entryHi] || this._tlb[this._entryHi | 0xFFF]) {
			return ;
		}

		// Clear out previous TLB element (if it was valid)
		if (this._tlbLo[index] & 0x200) {
			let indexWas = this._tlbHi[index] | ((this._tlbLo[index] & 0x100) ? 0xFFF : 0);
			delete this._tlb[indexWas];
		}

		// Setup our fast lookup
		if (this._entryLo & 0x200) {
			let indexIs = this._entryHi | ((this._entryLo & 0x100) ? 0xFFF : 0);
			this._tlb[indexIs]  = this._entryLo | index;
		}

		// Store our TLB (does not handle global)
		this._tlbLo[index] = this._entryLo;
		this._tlbHi[index] = this._entryHi;
	}

	_tlbr(pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		this._entryLo = this._tlbLo[(this._index >> 8) & 0x3F];
		this._entryHi = this._tlbHi[(this._index >> 8) & 0x3F];
	}

	_tlbp(pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		let found = this._tbl[this._entryHi] || this._tlb[this._entryHi | 0xFFF];

		if (found & 0x200) {
			this._index = (found & 0x3F) << 8;
		} else {
			this._index |= 0x80000000;
		}
	}

	_tlbwi(pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		this._writeTLB((this._index >> 8) & 0x3F, pc, delayed);
	}

	_tlbwr(pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		this._writeTLB(random(), pc, delayed);
	}

	// Exception logic
	_trap(e) {
		// TODO: Trap to COP0
		throw e;
	}

	_rfe(pc, delayed) {
		if (!this._copEnabled(0)) {
			throw new Exception(Const.Exceptions.CoprocessorUnusable, pc, delayed);
		}

		this._status = (this._status & ~0xF) | ((this._status >> 2) & 0xF);
	}
}

File: core/mips/cop0.cc
#include <stdint.h>

#include "compiler.h"
#include "fields.h"
#include "consts.h"
#include "cop0.h"

#include "imports.h"
#include "memory.h"

#include "registers.h"
#include "system.h"

static uint32_t status;

static uint32_t cause;
static uint32_t epc;
static uint32_t bad_addr;

static uint32_t page_table_addr;
static uint32_t process_state;

void COP0::reset() {
	status = STATUS_KUc | STATUS_BEV;
	cause = 0;
}

static inline uint32_t random() {
	static uint32_t inc = 0x8;

	if (++inc >= 0x40) inc = 0x8;

	return inc;
}

int cop_enabled(int cop) {
	return (((status >> 18) | ((status & STATUS_KUc) ? 1 : 0)) >> cop) & 1;
}

void COP0::interrupt(SystemIRQ i) {
	status |= (1 << (i + 8)) & STATUS_IM;
}

void COP0::handle_interrupt() {
	if ((status & STATUS_IEc) && (cause & status & STATUS_IM)) {
		trap(EXCEPTION_INTERRUPT, registers.pc, 0, 0);
	}
}

void COP0::bus_fault(int ex, uint32_t address, uint32_t pc, uint32_t delayed) {
	bad_addr = address;
	exception(ex, pc, delayed, 0);
}

uint32_t COP0::lookup(uint32_t address, bool write, bool& failure) {
	bool translated = (address & 0xC0000000) != 0xC0000000;
	bool is_kernel = status & STATUS_KUc;

	// Early supervisor test
	if ((address & 0x80000000) && !is_kernel) {
		failure = true;
		return ~0;
	}

	// Is address translated
	if (translated) {
		if (!cop_enabled(0)) {
			failure = true;
			return ~0;
		}

		uint32_t page_ptr = page_table_addr;
		int bits = 32;

		for (;;) {
			// Current page is supervisor only
			if ((page_ptr & PAGETABLE_KERNAL_MASK) && !is_kernel) {
				failure = true;
				return ~0;
			}

			// Write only page
			if ((page_ptr & PAGETABLE_RO_MASK) && write) {
				failure = true;
				return ~0;
			}

			// Not a global page (or TBL global disabled)
			if (PAGETABLE_GLOBAL_MASK & ~(page_ptr | process_state)) {
				if ((PAGETABLE_PID_MASK & process_state) != (PAGETABLE_PID_MASK & page_ptr)) {
					failure = true;
					return ~0;
				}
			}

			int length = (page_ptr & PAGETABLE_LEN_MASK) * 8;

			if (length == 0) {
				uint32_t mask = ~0 >> bits;
				return (address & mask) | (page_ptr & ~mask);
			}

			bits -= length;

			// Page table grainularity fault
			if (bits < 12) {
				failure = true;
				return ~0;
			}

			int index = (address >> bits) & ((1 << length) - 1);

			// Chain page table lookups (with a double fault check)
			page_ptr = Memory::read((page_ptr & PAGETABLE_ADDR_MASK) + (index * 4), failure);

			if (failure) return ~0;
		} 
	}
	
	// Unmapped through TLB
	return address & 0x1FFFFFFF;
}

EXPORT uint32_t translate(uint32_t address, uint32_t write, uint32_t pc, uint32_t delayed) {
	bool translated = (address & 0xC0000000) != 0xC0000000;
	bool is_kernel = status & STATUS_KUc;

	// Early supervisor test
	if ((address & 0x80000000) && !is_kernel) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, pc, delayed);
	}

	// Is address translated
	if (translated) {
		if (!cop_enabled(0)) {
			exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
		}

		uint32_t page_ptr = page_table_addr;
		int bits = 32;

		for (;;) {
			// Current page is supervisor only
			if ((page_ptr & PAGETABLE_KERNAL_MASK) && !is_kernel) {
				exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);			
			}

			// Write only page
			if ((page_ptr & PAGETABLE_RO_MASK) && write) {
				COP0::bus_fault(EXCEPTION_TLBMOD, address, pc, delayed);
			}

			// Not a global page (or TBL global disabled)
			if (PAGETABLE_GLOBAL_MASK & ~(page_ptr | process_state)) {
				if ((PAGETABLE_PID_MASK & process_state) != (PAGETABLE_PID_MASK & page_ptr)) {
					COP0::bus_fault(write ? EXCEPTION_TLBSTORE : EXCEPTION_TLBLOAD, address, pc, delayed);	
				}
			}

			int length = (page_ptr & PAGETABLE_LEN_MASK) * 8;

			if (length == 0) {
				uint32_t mask = ~0 >> bits;
				return (address & mask) | (page_ptr & ~mask);
			}

			bits -= length;

			// Page table grainularity fault
			if (bits < 12) {
				COP0::bus_fault(write ? EXCEPTION_TLBSTORE : EXCEPTION_TLBLOAD, address, pc, delayed);
			}

			int index = (address >> bits) & ((1 << length) - 1);

			// Chain page table lookups (with a double fault check)
			bool exception = false;
			page_ptr = Memory::read((page_ptr & PAGETABLE_ADDR_MASK) + (index * 4), exception);

			if (exception) {
				COP0::bus_fault(EXCEPTION_DOUBLEFAULT, address, pc, delayed);
			}
		} 
	}
	
	// Unmapped through TLB
	return address & 0x1FFFFFFF;
}

EXPORT void trap(int exception, int address, int delayed, int coprocessor) {
	// Preserve return address
	epc = address;

	// Enter kernal mode, and shift the mode flags
	status = (status & ~0x3F) | ((status << 2) & 0x3C) | STATUS_KUc;

	// Set our cause register
	cause = (cause & 0x0000FF00) |
		(delayed ? 0x80000000 : 0) |
		(coprocessor << 28) |
		(exception << 2);

	switch (exception) {
	case EXCEPTION_TLBLOAD:
	case EXCEPTION_TLBSTORE:
		registers.pc = (status & STATUS_BEV) ? TLB_EXCEPTION_VECTOR : REMAPPED_TLB_EXCEPTION_VECTOR;
		break ;

	default:
		registers.pc = (status & STATUS_BEV) ? EXCEPTION_VECTOR : REMAPPED_EXCEPTION_VECTOR;
		break ;
	}
}

// ******
// ** Co-Processor Move registers
// ******

EXPORT void MFC0(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	uint32_t value;

	switch (FIELD_RD(word)) {
	// Random registers
	case 0xF: // c0_prid
		value = PROCESSOR_ID;
		break ;

	// Virtual-memory registers
	case 0x0:
		value = page_table_addr;
		break ;
	case 0x1: // c0_random (non-deterministic, cheap method)
		value = random() << 8;
		break ;
	case 0x2:
		value = process_state;
		break ;
	case 0x8: // c0_vaddr
		value = bad_addr;
		break ;

	// Status/Exception registers
	case 0xC: // c0_status
		value = status;
		break ;
	case 0xD: // c0_cause
		value = cause;
		break ;
	case 0xE: // c0_epc
		value = epc;
		break ;
	default:
		value = 0;
		break ;
	}

	write_reg(FIELD_RT(word), value);
}

EXPORT void MTC0(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	uint32_t value = read_reg(FIELD_RT(word));

	switch (FIELD_RD(word)) {
	// Virtual address registers
	case 0x0:
		page_table_addr = value;
		break ;
	case 0x2:
		process_state = value & (PAGETABLE_GLOBAL_MASK | PAGETABLE_PID_MASK);
		break ;

	// Status/Exception registers
	case 0xC: // c0_status
		status = value & ALL_STATUS_BITS;
		break ;
	case 0xD: // c0_cause
		cause = (cause & ~0x0000FF00) | (value & 0x0000FF00);
		break ;
	}
}

// ******
// ** Co-Processor instructions
// ******

EXPORT void RFE(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	status = (status & ~0xF) | ((status >> 2) & 0xF);
}

/*
EXPORT void TLBR(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	entry_lo = tlb_lo[(index >> 8) & 0x3F];
	entry_hi = tlb_hi[(index >> 8) & 0x3F];
}

EXPORT void TLBWI(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	write_tlb((index >> 8) & 0x3F);
}

EXPORT void TLBWR(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	write_tlb(random());
}

EXPORT void TLBP(uint32_t address, uint32_t word, uint32_t delayed) {
	if (!cop_enabled(0)) {
		exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
	}

	uint32_t found = read_tlb(entry_hi);

	if (found & 0x200) {
		index = (found & 0x3F) << 8;
	} else {
		index |= 0x80000000;
	}
}
*/

// ***********
// ** Unused move instructions
// ***********
EXPORT void CFC0(uint32_t address, uint32_t word, uint32_t delayed) {
	exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
}

EXPORT void CTC0(uint32_t address, uint32_t word, uint32_t delayed) {
	exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
}

EXPORT void LWC0(uint32_t address, uint32_t word, uint32_t delayed) {
	exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
}

EXPORT void SWC0(uint32_t address, uint32_t word, uint32_t delayed) {
	exception(EXCEPTION_COPROCESSORUNUSABLE, address, delayed, 0);
}
